### ISSUE TITLE
maven settings credential did not work

### ISSUE THREAD

**Issue by zhaofengcumt** [GoogleContainerTools/jib#545]
**Description of the issue**:

**Expected behavior**:

**Steps to reproduce**: 

**Environment**: 

**`jib-maven-plugin` Configuration**: 
```xml
            <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>0.9.2</version>
                <configuration>
                    <allowInsecureRegistries>true</allowInsecureRegistries>
                    <from>
                        <image>192.168.156.116/zhaofeng/nginx</image>
                    </from>
                    <to>
                        <image>192.168.156.116/zhaofeng/demo:8</image>
                    </to>
                </configuration>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>build</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

maven:
```
    <server>
      <id>192.168.156.116</id>
      <username>admin</username>
      <password>zf1010@QQ.com</password>
    </server>
```

**`jib-gradle-plugin` Configuration**: 
PASTE YOUR build.gradle CONFIGURATION HERE

**Log output**: 
 Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.2:build (default-cli) on project k8s-webapp-demo: Build image failed, perhaps you should make sure your credentials for '192.168.156.116' are set up correctly: com.google.api.client.http.HttpResponseException: 401 Unauthorized
[ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"zhaofeng/nginx","Action":"pull"}]}]}

**Additional Information**:

**Comment by coollog** [comment]
This is most likely related to #541 and will be fixed with #549 . The fix will be released in version 0.9.4 soon.

**Comment by coollog** [comment]
@zhaofengcumt We just released version `0.9.4`. Try updating to that version and your issue should hopefully be fixed :)

**Comment by zhaofengcumt** [comment]
it still failed.....

**Comment by coollog** [comment]
@zhaofengcumt Is it still the same error?

**Comment by zhaofengcumt** [comment]
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.5:build (default) on project k8s-webapp-demo: Build image failed, perhaps you should make sure your credentials for '192.168.156.116' are set up correctly: Unauthorized for 192.168.156.116/zhaofeng/nginx: 401 Unauthorized
[ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"zhaofeng/nginx","Action":"pull"}]}]}


<server>
      <id>192.168.156.116</id>
      <username>admin</username>
      <password>zf1010@QQ.com</password>
    </server>

**Comment by loosebazooka** [comment]
Can you provide a little information on how you set up your registries?

**Comment by zhaofengcumt** [comment]
this is my maven settings
<server>
      <id>192.168.156.116</id>
      <username>admin</username>
      <password>zf1010@QQ.com</password>
    </server>

this is the plugin
 <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>0.9.5</version>
                <configuration>
                    <allowInsecureRegistries>true</allowInsecureRegistries>
                    <from>
                        <image>192.168.156.116/zhaofeng/nginx</image>
                    </from>
                    <to>
                        <image>192.168.156.116/zhaofeng/demo:8</image>
                    </to>
                </configuration>

with the command docker login 192.168.156.116 i can success

**Comment by zhaofengcumt** [comment]
this is my maven settings
`
<server> <id>192.168.156.116</id> <username>admin</username> <password>zf1010@QQ.com</password> </server>
`



with the command docker login 192.168.156.116 i can success

**Comment by zhaofengcumt** [comment]
`      <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>0.9.5</version>
                <configuration>
                    <allowInsecureRegistries>true</allowInsecureRegistries>
                    <from>
                        <image>192.168.156.116/zhaofeng/nginx</image>
                    </from>
                    <to>
                        <image>192.168.156.116/zhaofeng/demo:8</image>
                    </to>
                </configuration>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>build</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>`

**Comment by coollog** [comment]
@zhaofengcumt , could you run `curl -v 192.168.156.116/v2/` and provide us the output of that?

**Comment by zhaofengcumt** [comment]
` curl -v 192.168.156.116/v2/
* About to connect() to 192.168.156.116 port 80 (#0)
*   Trying 192.168.156.116...
* Connected to 192.168.156.116 (192.168.156.116) port 80 (#0)
> GET /v2/ HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 192.168.156.116
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< Server: nginx
< Date: Thu, 12 Jul 2018 01:49:10 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 87
< Connection: keep-alive
< Docker-Distribution-Api-Version: registry/2.0
< Set-Cookie: beegosessionID=b42a43f9be0e2d102054fd2dbe5672e1; Path=/; HttpOnly
< Www-Authenticate: Bearer realm="http://192.168.156.116/service/token",service="harbor-registry"
< 
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
* Connection #0 to host 192.168.156.116 left intact`

**Comment by coollog** [comment]
Okay, great, now try: `curl -v  "192.168.156.116/service/token?service=harbor-registry&scope=repository:zhaofeng/nginx:pull" -u "admin:zf1010@QQ.com"`

By the way, did you happen to see a `Failed to retrieve authentication challenge for registry` in your jib:build logs?

**Comment by zhaofengcumt** [comment]
```
[root@udmkf3 ~]# curl -v "192.168.156.116/service/token?service=harbor-registry&scope=repository:zhaofeng/nginx:pull" -u "admin:zf1010@QQ.com"
* About to connect() to 192.168.156.116 port 80 (#0)
*   Trying 192.168.156.116...
* Connected to 192.168.156.116 (192.168.156.116) port 80 (#0)
* Server auth using Basic with user 'admin'
> GET /service/token?service=harbor-registry&scope=repository:zhaofeng/nginx:pull HTTP/1.1
> Authorization: Basic YWRtaW46emYxMDEwQFFRLmNvbQ==
> User-Agent: curl/7.29.0
> Host: 192.168.156.116
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx
< Date: Thu, 12 Jul 2018 02:15:26 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 1197
< Connection: keep-alive
< Set-Cookie: beegosessionID=990fd01ee8139652f67895b620e369d5; Path=/; HttpOnly
< 
{
  "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IkdTUFk6QTVQMjpLNzNNOkJLRlg6VkEzSDpWQlBBOjJQS1E6MzdBSzo0VDRBOkZWSjY6S0s1RDpOUERaIn0.eyJpc3MiOiJoYXJib3ItdG9rZW4taXNzdWVyIiwic3ViIjoiYWRtaW4iLCJhdWQiOiJoYXJib3ItcmVnaXN0cnkiLCJleHAiOjE1MzEzNjM1MjYsIm5iZiI6MTUzMTM2MTcyNiwiaWF0IjoxNTMxMzYxNzI2LCJqdGkiOiJ1TWUzM21IdzE5SjJDUGVZIiwiYWNjZXNzIjpbeyJ0eXBlIjoicmVwb3NpdG9yeSIsIm5hbWUiOiJ6aGFvZmVuZy9uZ2lueCIsImFjdGlvbnMiOlsicHVzaCIsIioiLCJwdWxsIl19XX0.C7Gbj_ZqIFOVk1x9VxVbOf6L0uNWY_0OwheV9EwKweWYN0J9rLqXxFTm_RgaM8RxAKLsvVTEdBziF89ZjwIyZ4F0gNPNeuPkiBwdARNXpCnxRTMwNlY6TG3l-QXzRN9STGbEtbuRXncOUEnoxVzUuExzr7gxzQzAVxKNEDq4x81pwDVFYeaLZjGw9y-Hon_kXjYHPXawsLRYrceN6a_AGZ2I0Q63XWvsekpkvJevMDXdiUvJq2NZCy38DSpOE-HeImtDZvm7cfdNYE_Tzd3pVxTg2Kp7Y41zqzFBNmu9Ea_hR_vkq6iCpJ_XFqwHMenm4UJTFjr_324cG1SHBhMDBpnzthICBkluT71brPiNiI6H10Nk1fRHl8w4rsIWvgWGCoQib93DEb3-KzHodNvP-xWxy98f5jlDxZRDh4vKw36wt998oRoUixnFEyNkpO8H4Terc_ZFVOdg4iLjBG4XGr8hrUw9rNml_1FtCejSgRQa-PIBkr3xAkKki3lvptuGdvCxt9C16CgJhbLv-LhXw7nazJ7kmJAeNtcJi4SvBbobnOFGB3F-12LVUtlL4Q_AjqZLvPjOfSkSZjrmSKz-N5s1wtU1TO-IbkSzP4Ukkv2pbD_YCOOijDi0uS6Ktmwv-P82oY36HqCZriLN987hf8jDExIPRUW-_nygdN_iE0s",
  "expires_in": 1800,
  "issued_at": "2018-07-12T02:15:26Z"
* Connection #0 to host 192.168.156.116 left intact
}[root@udmkf3 ~]# 







```

```
i did see Failed to retrieve authentication challenge for registry
[INFO] Building resources layer...
[INFO] Building classes layer...
[INFO] Retrieving registry credentials for 192.168.156.116...
[ERROR] Failed to retrieve authentication challenge for registry that required token authentication


```

**Comment by coollog** [comment]
Ah okay, it looks like we are not giving the right error message for your failure case. To solve your issue, you need to add an `<allowInsecureRegistries>true</allowInsecureRegistries>` to your `jib-maven-plugin` configuration (see [Extended Usage](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#extended-usage)).

**Comment by coollog** [comment]
Ah, you're right, I somehow missed that. We'll keep investigating to resolve your issue.

**Comment by coollog** [comment]
Looks like we have a bug in that we aren't allowing HTTP registries to do zero-credential token authentication even with `allowInsecureRegistries` set to `true`.

**Comment by chanseokoh** [comment]
@coollog if `allowInsecureRegistries` is set to `true`, does Jib stick to HTTP from the beginning or it can fall back to HTTP only after the HTTPS connection is not properly established?

**Comment by chanseokoh** [comment]
It seems worth printing out the message if the fallback happens.

**Comment by coollog** [comment]
@zhaofengcumt Nope, I will update this thread when it is ready. Most likely will be sometime in the morning (EST).
@chanseokoh +1

**Comment by coollog** [comment]
Hi @zhaofengcumt we actually just released version `0.9.6` - let us know if that fixes your issue :)

**Comment by zhaofengcumt** [comment]
```
[INFO] Retrieving registry credentials for 192.168.156.116...
[INFO] Getting base image 192.168.156.116/zhaofeng/nginx:v1.10.3...
[INFO] Building dependencies layer...
[INFO] Building resources layer...
[INFO] Building classes layer...
[INFO] Retrieving registry credentials for 192.168.156.116...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.363 s
[INFO] Finished at: 2018-07-12T12:27:04+08:00
[INFO] Final Memory: 38M/462M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:build (default) on project k8s-webapp-demo: Build image failed, perhaps you should make sure your credentials for '192.168.156.116' are set up correctly: Unauthorized for 192.168.156.116/zhaofeng/nginx: 401 Unauthorized
[ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"zhaofeng/nginx","Action":"pull"}]}]}
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
```

**Comment by coollog** [comment]
Ah, can you update that comment with the build output when running with the `-X` flag? (like `mvn compile jib:build -X`)

**Comment by coollog** [comment]
@zhaofengcumt I think you posted to the wrong issue haha (https://github.com/GoogleContainerTools/jib/issues/542#issuecomment-404388548). But I looked at that and I realized that you were trying to use your registry with credentials over HTTP, which we don't actually support right now (you can only use credentials with registries that communicate via HTTPS). We can possibly discuss what the best way to approach this is, but for now, I would recommend setting up HTTPS on your private registry.

**Comment by zhaofengcumt** [comment]
https? if so the allowInsecureRegistries  is used for what

**Comment by coollog** [comment]
`allowInsecureRegistries` allows for communication with registries over HTTP (by falling back from HTTPS if failed), but we explicitly chose not to send any authorization credentials even when provided due to a security concern. Basically, if you were to `allowInsecureRegistries` to be able to pull from a registry that used HTTP, and pushed to a registry that used HTTPS, you would expect the push to the registry that used HTTPS to not have to fall back to HTTP. However, when you send a request to the HTTPS registry, a malicious individual on your network could send a failure response. This would cause your request to be retried in HTTP. If that request contained your credentials, the malicious individual would be able to sniff that.

We can try and explore how to possibly:
1. Change our configuration to account be separate for each registry, or
1. Change `allowInsecureRegistries` to be a encompass sending credentials with a big security warning

**Comment by chanseokoh** [comment]
@zhaofengcumt  you will be able to send the password over HTTP (which can be seen by anyone in the network) using the new system property `sendCredentialsOverHttp` (#641) once Jib 0.9.7 is released.

**Comment by coollog** [comment]
Hi @zhaofengcumt , we have released version `0.9.7` that includes the new `sendCredentialsOverHttp` system property to allow for sending credentials over HTTP. Can you retry with the new version? Closing but feel free to reopen if issues persist.
